pkg/agent/apis/server.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apis
+
+import (
+	"fmt"
+	"net"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	k8sversion "k8s.io/apimachinery/pkg/version"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	genericoptions "k8s.io/apiserver/pkg/server/options"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/apis/handlers"
+	_ "github.com/vmware-tanzu/antrea/pkg/agent/apis/handlers/addressgroup"
+	_ "github.com/vmware-tanzu/antrea/pkg/agent/apis/handlers/agentinfo"
+	_ "github.com/vmware-tanzu/antrea/pkg/agent/apis/handlers/appliedtogroup"
+	_ "github.com/vmware-tanzu/antrea/pkg/agent/apis/handlers/rule"
+	"github.com/vmware-tanzu/antrea/pkg/monitor"
+	antreaversion "github.com/vmware-tanzu/antrea/pkg/version"
+)
+
+const (
+	Name = "antrea-agent-api"
+	Port = 10443
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+type agentAPIServer struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+}
+
+func (s *agentAPIServer) Run(stopCh <-chan struct{}) error {
+	return s.GenericAPIServer.PrepareRun().Run(stopCh)
+}
+
+func installHandlers(aq monitor.AgentQuerier, npq monitor.AgentNetworkPolicyInfoQuerier, s *genericapiserver.GenericAPIServer) {
+	q := handlers.Querier{
+		AgentQuerier:             aq,
+		NetworkPolicyInfoQuerier: npq,
+	}
+	r := handlers.Registry()
+	for p, h := range r {
+		s.Handler.NonGoRestfulMux.HandleFunc(p, h.HandleFunc(q))
```

Comment:
I feel the code can be simpler and more readable in this way:
1. Instead of relying on init function and side effect of unused imports and an extra Registry, install handlers with their path and their handlers explicitly, which can be clear that what path and handlers we have.
2. Instead of asking each handler to implement a generic interface which provides all queriers to all handlers, just use closure to pass necessary queriers to each handler.

For example:
```
s.Handler.NonGoRestfulMux.HandleFunc("/agentinfo", agentinfo.Handler(AgentQuerier))
s.Handler.NonGoRestfulMux.HandleFunc("/rules", rule.Handler(NetworkPolicyInfoQuerier))
s.Handler.NonGoRestfulMux.HandleFunc("/appliedtogroups", appliedtogroup.Handler(NetworkPolicyInfoQuerier))
s.Handler.NonGoRestfulMux.HandleFunc("/addressgroups", addressgroup.Handler(NetworkPolicyInfoQuerier))

# rule/handler.go can be simplified as:
func (h *Handler) Handler(networkPolicyInfoQuerier monitor.AgentNetworkPolicyInfoQuerier) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		rules := networkPolicyInfoQuerier.GetRules()
		if err := json.NewEncoder(w).Encode(rules); err != nil {
			http.Error(w, "Failed to encode response: "+err.Error(), http.StatusInternalServerError)
		}
	}
}
```

Reply:
Thanks, edited.